include/swift/SILOptimizer/Utils/DeinitBarrierFinding.h

```diff
@@ -0,0 +1,53 @@
+//===- DeinitBarrierFinding.h - Find deinit barriers from roots -*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// Given a set of roots, find the deinit barriers that exist between it and the
+// those instructions which the client says are already barriers (think users).
+//===----------------------------------------------------------------------===//
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+
+namespace swift {
+
+class SILBasicBlock;
+class SILFunction;
+class SILInstruction;
+class BasicCalleeAnalysis;
+
+using llvm::ArrayRef;
+using llvm::function_ref;
+
+namespace DeinitBarrierFinding {
+/// How destroy_value hoisting is obstructed.
+struct Barriers final {
+  /// Instructions which are deinit barriers.
+  llvm::SmallVector<SILInstruction *, 4> instructions;
+
+  /// Blocks one of whose phis is a deinit barrier.
+  llvm::SmallVector<SILBasicBlock *, 4> phis;
+
+  llvm::SmallVector<SILBasicBlock *, 4> blocks;
+
+  Barriers() {}
+  Barriers(Barriers const &) = delete;
+  Barriers &operator=(Barriers const &) = delete;
+};
+
+} // namespace DeinitBarrierFinding
+
+void findDeinitBarriersAboveInstructions(
```

Comment:
What makes this specific to deinit barriers? This looks like a simple adapter that converts Reachability's Visitor pattern to a simple `isBarrier` callback. That's pretty handy to avoid implementing a visitor. But can this utility just do that: take your `isBarrier`, wrap up a Visitor, and give you back a set of barriers?

The user needs to pass in a barrier function anyway, so can't they just pass in a function that takes a pointer to the Context and checks for deinit barriers?